src/peep/peep.c

```diff
@@ -8671,23 +8673,456 @@ static void peep_easter_egg_peep_interactions(rct_peep *peep)
 }
 
 /**
+ * rct2: 0x0069101A
+ *
+ * @return (CF)
+ */
+static bool sub_69101A(rct_map_element *esi) {
+	rct_ride *ride = get_ride(esi->properties.track.ride_index);
+	if (RCT2_ADDRESS(0x97C3AF, uint8)[ride->type] == 0) {
+		return true;
+	}
+
+	if (ride->excitement == 0xFFFF) {
+		return false;
+	}
+
+	if (ride->excitement > 0x1D6) {
+		return false;
+	}
+
+	if (ride->intensity > 0x1C2) {
+		return false;
+	}
+
+	if (RideData4[ride->type].flags & 0x2000) {
+		if ((scenario_rand() & 0xFFFF) > 0x3333) {
+			return true;
+		}
+	} else if (RideData4[ride->type].flags & 0x4000) {
+		if ((scenario_rand() & 0xFFFF) > 0x1000) {
+			return true;
+		}
+	} else {
+		return true;
+	}
+
+	return false;
+}
+
+/**
  *
  *  rct2: 0x00690B99
  */
-static bool sub_690B99(rct_peep *peep, int edge, uint8 *rideToView, uint8 *rideSeatToView)
+static bool new_sub_690B99(rct_peep *peep, uint8 edge, uint8 *rideToView, uint8 *rideSeatToView)
 {
+	// TODO: map_get_surface_element_at()
+	rct_map_element *esi_element = map_get_first_element_at(peep->next_x / 32, peep->next_y / 32);
+	while (map_element_get_type(esi_element) != MAP_ELEMENT_TYPE_SURFACE) {
+		esi_element++;
+	}
+
+	do {
+		// loc_690BC9:
+		if (map_element_get_type(esi_element) == MAP_ELEMENT_TYPE_FENCE) {
+			if (map_element_get_direction(esi_element) == edge) {
+				rct_scenery_entry *entry = g_wallSceneryEntries[esi_element->properties.fence.type];
+				if (!(entry->wall.flags2 & WALL_SCENERY_FLAG4)) {
+					if (esi_element->base_height > peep->next_z + 4) {
+						if (esi_element->clearance_height < peep->next_z + 1) {
+							// loc_690FB1
+							return false;
+						}
+					}
+				}
+			}
+		}
+	} while (!map_element_is_last_for_tile(esi_element++));
+
+	uint16 x = peep->next_x + TileDirectionDelta[edge].x;
+	uint16 y = peep->next_y + TileDirectionDelta[edge].y;
+	if (x > 0x1FFF || y > 0x1FFF) {
+		return false;
+	}
+
+	// TODO: map_get_surface_element_at()
+	esi_element = map_get_first_element_at(x / 32, y / 32);
+	while (map_element_get_type(esi_element) != MAP_ELEMENT_TYPE_SURFACE) {
+		esi_element++;
+	}
+
+	do {
+		if (map_element_get_type(esi_element) == MAP_ELEMENT_TYPE_FENCE) {
+			if ((map_element_get_direction(esi_element) ^ 0x2) == edge) {
+				rct_scenery_entry *entry = g_wallSceneryEntries[esi_element->properties.fence.type];
+				if (!(entry->wall.flags2 & WALL_SCENERY_FLAG4)) {
+					if (peep->next_z + 4 >= esi_element->base_height) {
+						if (peep->next_z < esi_element->clearance_height) {
+							return false;
+						}
+					}
+				}
+			}
+		}
+	} while (!map_element_is_last_for_tile(esi_element++));
+
+	// save x + y
+
+	// TODO: map_get_surface_element_at()
+	rct_map_element *esi_element_3 = map_get_first_element_at(x / 32, y / 32);
+	while (map_element_get_type(esi_element_3) != MAP_ELEMENT_TYPE_SURFACE) {
+		esi_element++;
```

Comment:
`esi_element3` but really you should just refactor this all so that you only access the surface element once.